torch/_functorch/aot_autograd.py

```diff
@@ -2036,6 +2094,293 @@ def wrapped_compiled_fn(args):
+                    #
+                    # Alternatively we could run functionalization and make_fx() directly here,
+                    # although keeping them separate feels cleaner.
+                    nonlocal subclass_meta
+                    if is_primals:
+                        assert subclass_meta.fw_output_infos is None
+                        subclass_meta.fw_output_infos = unwrap_infos
+                        assert subclass_meta.fw_output_types is None
+                        subclass_meta.fw_output_types = [type(x) for x in outs]
+                    else:
+                        assert subclass_meta.bw_output_infos is None
+                        subclass_meta.bw_output_infos = unwrap_infos
+                return curr_outs_unwrapped
+            if is_joint:
+                assert len(outs_wrapped) == 2
+                assert isinstance(outs_wrapped[0], (list, tuple))
+                assert isinstance(outs_wrapped[1], (list, tuple))
+                outs_unwrapped = (
+                    type(outs_wrapped[0])(unwrap_output_list(outs_wrapped[0], is_primals=True)),
+                    type(outs_wrapped[1])(unwrap_output_list(outs_wrapped[1], is_primals=False)),
+                )
+            else:
+                assert isinstance(outs_wrapped, list)
+                outs_unwrapped = list(unwrap_output_list(outs_wrapped, is_primals=True))
+            return outs_unwrapped
+        return wrapped_flat_fn
+
+    # I'm actually not convinced that we can easily "pre-compute" the updated fw_metadata
+    # without manually running the whole thing.
+    # This means that we're performing tracing a second time.
+    # The silver lining is that this only happens when you actual trace with a TorchDispatchMode
+    curr_modes = _get_current_dispatch_mode_stack()
+    # First, sanity assert. FakeTensorMode is used by tracing, and should be the most recent
+    # mode pushed onto the mode stack
+    assert len(curr_modes) > 0
+    assert type(curr_modes[-1]) == FakeTensorMode
+    # Next, user asserts: We only support a single TorchDispatchMode usage for now
+    assert len(curr_modes) == 2, "AOTDispatch only supports a single TorchDispatchMode for now. \
+        If you're doing something involving multiple modes, please file an issue!"
+    assert type(curr_modes[0]) != FakeTensorMode
+
+    # This is a bit questionable: in particular, we are assuming that the entire graph
+    # should be traced with the user mode enabled.
+    # In practice this might not be the case: the user mode might be disabled halfway through the function.
+    # Maybe this is where we'll need higher order functions
+    with _pop_mode_temporarily() as fake_mode, _pop_mode_temporarily() as user_mode:
+        user_mode._dispatch_key = torch._C.DispatchKey.Functionalize
+        # Currently, python dispatcher is required when using per-dispatch-key modes
+        with enable_python_dispatcher(), fake_mode, user_mode:
```

Comment:
Interested in takes on this. I'm temporarily removing the user's mode from the Python mode stack, and pushing it onto the Functionalize mode stack just during tracing. The whole thing is pretty delicate, so for now I have a bunch of asserts about the mode stack (only one user mode is allowed, bottom mode should be FakeTensorMode).

Reply:
One way this is definitely broken today is if:
(a) the user really is using a `TorchDispatchMode` (no subclasses involved).
(b) the graph dynamo sends us involves the context manager getting destructed somewhere in the middle of the graph.

The bad behavior is that we will blindly trace the entire function as if the mode was enabled, but it was meant to be disabled halfway through tracing. The easy fix (that's probably already the case today) is if dynamo graph breaks on context manager start/end.

Reply:
If we make fake and proxy modes "dispatch keys" that have a fixed location does that help?